cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw.cpp

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+inline auto max_wg(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::max_work_group_size>();
+}
+
+template <class Float, class BinaryOp, class UnaryOp>
+class kernel_reduction_rm_cw_inplace {
+    typedef const Float* inp_t;
+    typedef Float* out_t;
+
+public:
+    kernel_reduction_rm_cw_inplace(inp_t const input_,
+                                   out_t const output_,
+                                   const std::int32_t width_,
+                                   const std::int64_t height_,
+                                   const std::int32_t lstride_,
+                                   const BinaryOp binary_,
+                                   const UnaryOp unary_)
+            : input{ input_ },
+              output{ output_ },
+              unary{ unary_ },
+              binary{ binary_ },
+              width{ width_ },
+              height{ height_ },
+              lstride{ lstride_ } {}
+
+    void operator()(sycl::nd_item<2> it) const {
+        using sycl::ONEAPI::reduce;
+        // Common for whole WG
+        const auto col_idx = it.get_global_id(0);
+        const auto loc_idx = it.get_global_id(1);
+        const auto range = it.get_global_range(1);
+        // Exclusive for EU
+        Float acc = binary.init_value;
+        for (std::int64_t i = loc_idx; i < height; i += range) {
+            inp_t const inp_row = input + lstride * i;
+            acc = binary.native(acc, unary(inp_row[col_idx]));
+        }
+        // WG reduction
+        auto grp = it.get_group();
+        output[col_idx] = reduce(grp, acc, binary.native);
```

Comment:
Does it works fine for big groups?

Reply:
It seems to work well. At least with WG = 256 | 128